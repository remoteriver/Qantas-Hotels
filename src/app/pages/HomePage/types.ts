import { Color } from '@material-ui/core';
import { Hotel } from 'types/Hotel';

/* --- STATE --- */
export interface HomePageState {
  results: Hotel[];
}

export type ContainerState = HomePageState;

export interface Data {
  results: Hotel[];
}
