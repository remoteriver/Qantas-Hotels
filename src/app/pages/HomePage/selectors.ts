import { createSelector } from '@reduxjs/toolkit';

import { RootState } from 'types';

const select = (state: RootState) => state.homepageState;

export const selectResults = createSelector([select], state => state?.results);
