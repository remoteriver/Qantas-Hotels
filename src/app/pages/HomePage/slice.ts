import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ContainerState, Data } from './types';

export const initialState: ContainerState = {
  results: [],
};

const homepageSlice = createSlice({
  name: 'homepageState',
  initialState,
  reducers: {
    getData(state) {},
    getDataSuccess(state, action: PayloadAction<Data>) {
      state.results = action.payload.results;
    },
  },
});

export const { actions, reducer, name: sliceKey } = homepageSlice;
