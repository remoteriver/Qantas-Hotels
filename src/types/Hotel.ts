export type Hotel = {
  id: string;
  property: Property;
  offer: Offer;
};

export type Property = {
  propertyId: string;
  title: string;
  address: string[];
  previewImage: PreviewImage;
  rating: Rating;
};

export type PreviewImage = {
  url: string;
  caption: string;
  imageType: ImageType;
};

export enum ImageType {
  PRIMARY = 'PRIMARY',
}

export type Rating = {
  ratingValue: number;
  ratingType: RatingType;
};

export enum RatingType {
  SELF = 'self',
  STAR = 'STAR',
}

export type Offer = {
  promotion: Promotion;
  name: string;
  displayPrice: Price;
  savings?: Price;
  cancellationOption: CancellationOption;
};

export type Promotion = {
  title: string;
  type: PromotionType;
};

export enum PromotionType {
  MEMBER = 'MEMBER',
}

export type Price = {
  amount: number;
  currency: Currency;
};

export enum Currency {
  AUD = 'AUD',
}

export type CancellationOption = {
  cancellationType: CancellationType;
};

export enum CancellationType {
  FREE_CANCELLATION = 'FREE_CANCELLATION',
  NOT_REFUNDABLE = 'NOT_REFUNDABLE',
}

/**
 * "id": "cxd650nuyo",
      "property": {
        "propertyId": "P107801",
        "title": "Courtyard by Marriott Sydney-North Ryde",
        "address": [
          "7-11 Talavera Rd",
          "North Ryde"
        ],
        "previewImage": {
          "url": "https://unsplash.it/145/125/?random",
          "caption": "Image of Courtyard by Marriott Sydney-North Ryde",
          "imageType": "PRIMARY"
        },
        "rating": {
          "ratingValue": 4.5,
          "ratingType": "self"
        }
      },
      "offer": {
        "promotion": {
          "title": "Exclusive Deal",
          "type": "MEMBER"
        },
        "name": "Deluxe Balcony Room",
        "displayPrice": {
          "amount": 329.000000000,
          "currency": "AUD"
        },
        "savings": {
          "amount": 30.000000000,
          "currency": "AUD"
        },
        "cancellationOption": {
          "cancellationType": "NOT_REFUNDABLE"
        }
      }
 */
