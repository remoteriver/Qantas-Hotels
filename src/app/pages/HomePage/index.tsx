import { Box, Container } from '@material-ui/core';
import * as React from 'react';
import { Helmet } from 'react-helmet-async';
import { useDispatch, useSelector } from 'react-redux';
import { useInjectReducer, useInjectSaga } from 'utils/redux-injectors';
import { useEffectOnMount } from 'utils/useEffectOnMount';
import { Header } from './components/Header';
import { HotelList } from './components/HotelList';
import { saga } from './saga';
import { selectResults } from './selectors';
import { sliceKey, reducer, actions } from './slice';

export const HomePage = () => {
  useInjectReducer({ key: sliceKey, reducer: reducer });
  useInjectSaga({ key: sliceKey, saga: saga });

  const dispatch = useDispatch();
  const results = useSelector(selectResults);

  useEffectOnMount(() => {
    dispatch(actions.getData());
  });

  return (
    <>
      <Helmet>
        <title>Home Page</title>
        <meta name="description" content="A Boilerplate application homepage" />
      </Helmet>
      <Container maxWidth="md">
        <Box p={2}>
          <Header></Header>
          <HotelList hotels={results}></HotelList>
        </Box>
      </Container>
    </>
  );
};
