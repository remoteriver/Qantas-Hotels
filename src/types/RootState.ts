// [IMPORT NEW CONTAINERSTATE ABOVE] < Needed for generating containers seamlessly

import { HomePageState } from 'app/pages/HomePage/types';

/* 
  Because the redux-injectors injects your reducers asynchronously somewhere in your code
  You have to declare them here manually
*/
export interface RootState {
  homepageState?: HomePageState;
  // [INSERT NEW REDUCER KEY ABOVE] < Needed for generating containers seamlessly
}
