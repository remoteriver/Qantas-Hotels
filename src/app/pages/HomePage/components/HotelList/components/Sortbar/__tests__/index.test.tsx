import { render } from '@testing-library/react';
import { SortBy } from 'app/pages/HomePage/types';
import { createRenderer } from 'react-test-renderer/shallow';
import { Sortbar } from '..';

describe('<HotelCard />', () => {
  beforeAll(() => {});

  const mockFunction = jest.fn();

  it('should render and match the snapshot', () => {
    //act
    const component = render(
      <Sortbar
        sortby={SortBy.PRICE_LOW_HIGH}
        handleSort={mockFunction}
      ></Sortbar>,
    );

    //assert
    expect(component.baseElement).toMatchSnapshot();
  });
});
