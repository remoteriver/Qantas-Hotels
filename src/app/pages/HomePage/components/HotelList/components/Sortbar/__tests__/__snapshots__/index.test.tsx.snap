// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HotelCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-3"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        id="sort-select-label"
      >
        Sort by
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="sort-select-label sort-select"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="sort-select"
          role="button"
          tabindex="0"
        >
          Price (low-high)
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="0"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</body>
`;
